File: src/shared/Systems/UpdateModelAttribute.lua
local RunService = game:GetService("RunService")
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local Components = require(ReplicatedStorage.Source.Components)

type ModelComponent = Components.ModelComponent

local name = RunService:IsServer() and "serverEntityId" or "clientEntityId"

--[=[
    @server @client

    Adds the id of an entity to a model for reconciliation purposes.
]=]
local function updateModelAttribute(world)
    for id: number, modelRecord in world:queryChanged(Components.Model) do
        local model: ModelComponent = modelRecord.new
        if model then
            model.model:SetAttribute(name, id)
        end
    end
end

return updateModelAttribute

File: src/shared/systems/update-model-attribute.ts
import { World } from "@rbxts/matter";
import { RunService } from "@rbxts/services";
import * as Components from "shared/components";

const name = RunService.IsServer() ? "serverEntityId" : "clientEntityId";

/**
 * Adds entity id to renderable model in the game.
 *
 * @param world Matter
 */
function updateModelAttribute(world: World): void {
	for (const [id, modelRecord] of world.queryChanged(Components.Renderable)) {
		const model = modelRecord.new;
		if (model) {
			model.model.SetAttribute(name, id);
		}
	}
}

export = updateModelAttribute;
